File: Social/Chat/VoiceRecorder.swift
//
//  VoiceRecorder.swift
//  Social
//
//  Grabación de voz nativa (`AVAudioRecorder`, sin SDK de terceros) —
//  última pieza real de "chat funcional con fotos, voz, reacciones, read
//  receipts". AAC en un contenedor .m4a. Equivalente de VoiceRecorder.kt.
//
//  Aviso de honestidad: configuración estándar de AVAudioSession/
//  AVAudioRecorder documentada por Apple, coherente con el resto del
//  proyecto (misma disciplina que AVFoundation ya usado en la cámara) —
//  sin verificación de compilador real (límite de plataforma).
//

import Foundation
import AVFoundation

final class VoiceRecorder: NSObject {
    private var recorder: AVAudioRecorder?
    private var outputURL: URL?

    func start() throws -> URL {
        let session = AVAudioSession.sharedInstance()
        try session.setCategory(.playAndRecord, mode: .default)
        try session.setActive(true)

        let url = FileManager.default.temporaryDirectory.appendingPathComponent("voice_\(UUID().uuidString).m4a")
        outputURL = url

        let settings: [String: Any] = [
            AVFormatIDKey: Int(kAudioFormatMPEG4AAC),
            AVSampleRateKey: 44100,
            AVNumberOfChannelsKey: 1,
            AVEncoderAudioQualityKey: AVAudioQuality.medium.rawValue
        ]
        let r = try AVAudioRecorder(url: url, settings: settings)
        r.record()
        recorder = r
        return url
    }

    /// Devuelve el archivo grabado, o nil si no había grabación en curso o
    /// fue demasiado corta para tener audio útil.
    func stop() -> URL? {
        recorder?.stop()
        recorder = nil
        return outputURL
    }

    func cancel() {
        recorder?.stop()
        recorder = nil
        if let outputURL {
            try? FileManager.default.removeItem(at: outputURL)
        }
        outputURL = nil
    }
}

File: Android/app/src/main/java/com/social/app/chat/VoiceRecorder.kt
package com.social.app.chat

import android.content.Context
import android.media.MediaRecorder
import android.os.Build
import java.io.File

/**
 * Grabación de voz nativa (`MediaRecorder`, sin SDK de terceros) — última
 * pieza real de "chat funcional con fotos, voz, reacciones, read
 * receipts". AAC en un contenedor MPEG_4 (.m4a), formato estándar
 * reproducible por `MediaPlayer` sin conversión.
 */
class VoiceRecorder(private val context: Context) {

    private var recorder: MediaRecorder? = null
    private var outputFile: File? = null

    fun start(): File {
        val file = File.createTempFile("voice_", ".m4a", context.cacheDir)
        outputFile = file
        @Suppress("DEPRECATION")
        val r = if (Build.VERSION.SDK_INT >= 31) MediaRecorder(context) else MediaRecorder()
        r.setAudioSource(MediaRecorder.AudioSource.MIC)
        r.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
        r.setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
        r.setOutputFile(file.absolutePath)
        r.prepare()
        r.start()
        recorder = r
        return file
    }

    /** Devuelve el archivo grabado, o null si no había grabación en curso. */
    fun stop(): File? {
        try {
            recorder?.stop()
        } catch (e: Exception) {
            // stop() puede lanzar si se para casi inmediatamente después de
            // start() (grabación demasiado corta) — no hay audio útil que
            // subir en ese caso.
            return null
        } finally {
            recorder?.release()
            recorder = null
        }
        return outputFile
    }

    fun cancel() {
        try {
            recorder?.stop()
        } catch (e: Exception) {
            // Ignorado — se está descartando de todas formas.
        }
        recorder?.release()
        recorder = null
        outputFile?.delete()
        outputFile = null
    }
}
